src/darwin/Framework/CHIP/MTRDevice_XPC.mm

```diff
@@ -188,16 +190,48 @@ - (oneway void)deviceConfigurationChanged:(NSNumber *)nodeID
 
 - (oneway void)device:(NSNumber *)nodeID internalStateUpdated:(NSDictionary *)dictionary
 {
+    MTR_LOG("%@ %s", self, __PRETTY_FUNCTION__);
+
+    // Save old state for comparison later
+    NSNumber * oldStateNumber = self._internalState[kMTRDeviceInternalPropertyDeviceState];
+    NSNumber * newStateNumber = dictionary[kMTRDeviceInternalPropertyDeviceState];
     [self _setInternalState:dictionary];
-    MTR_LOG("%@ internal state updated", self);
+
+    // Call delegate if state changed. State is considered changed if:
+    // 1) old state is nil but new state is not nil
+    // 2) old state is not nil but new state is nil
+    // 3) both old and new state are not nil, and they are not equal
+    if ((!oldStateNumber && newStateNumber) || (oldStateNumber && !newStateNumber) || (oldStateNumber && newStateNumber && ![newStateNumber isEqualToNumber:oldStateNumber])) {
+        MTRDeviceState state = static_cast<MTRDeviceState>(newStateNumber ? newStateNumber.unsignedIntegerValue : MTRDeviceStateUnknown);
+        [self _lockAndCallDelegatesWithBlock:^(id<MTRDeviceDelegate> delegate) {
+            [delegate device:self stateChanged:state];
+        }];
+    }
 }
 
-#pragma mark - Remote Commands
+- (MTRDeviceState)state
+{
+    NSNumber * stateNumber = self._internalState[kMTRDeviceInternalPropertyDeviceState];
```

Comment:
I don't see us allocating any _internalState on init, so it might be nil.  What does this square bracket dictionary notation do when the dictionary is nil?  Just return nil?

Reply:
It does